### DISCUSSION TITLE
Copied value of indexed casadi array leads to an error in MPC setup

### DISCUSSION THREAD

**Discussion by 19MTG04** [do-mpc/do-mpc#404] [Q&A]
I would like to use a part of a model state in my objective function. Therefore I try to reference a part of the state vector **phi** with my variable **a** (which will copy and not reference it). The mpc.setup() will fail if I create **a** before the model setup, but it works if I create it afterwards. 
Why is that the case and is there a possibility to actually store a reference of the state vector index in a variable before the model setup?

``` python
import do_mpc

model_type = 'continuous'
model = do_mpc.model.Model(model_type)

phi = model.set_variable(var_type='_x', var_name='phi', shape=(2, 1))
input = model.set_variable(var_type='_u', var_name='input')
model.set_rhs('phi', input)

a = phi[0]
model.setup()

# This would make the objective function work
# a = phi[0]

mpc = do_mpc.controller.MPC(model)
setup_mpc = {'n_horizon': 20, 't_step': 0.1}
mpc.set_param(**setup_mpc)

mterm = a**2
lterm = a**2
mpc.set_objective(mterm=mterm, lterm=lterm)

mpc.setup()

```

**Which version of do-mpc are using and how did you obtain it?**
- Installation via pip
- v4.4.0

**Which system are you working on?**
 - Windows

**Comment by 4flixt** [comment]
Hi,

apologies for the late reply. I feel like this is still relevant, however. We suggest the following approach:

```python
import do_mpc

model_type = 'continuous'
model = do_mpc.model.Model(model_type)

phi = model.set_variable(var_type='_x', var_name='phi', shape=(2, 1))
input = model.set_variable(var_type='_u', var_name='input')
model.set_rhs('phi', input)

model.setup()

mpc = do_mpc.controller.MPC(model)
setup_mpc = {'n_horizon': 20, 't_step': 0.1}
mpc.set_param(**setup_mpc)

mterm = mpc.x['a']**2
lterm = mpc.x['a']**2
mpc.set_objective(mterm=mterm, lterm=lterm)

mpc.setup()
```

The core idea is, that each module (MPC, MHE, Simulator) get their own copy of the model variables. Those should be used for the configuration of the respective model. 

Best,
4flixt
